sycl/doc/extensions/Reduction/Reduction_status.md

```diff
@@ -0,0 +1,189 @@
+    typename Reduction::reducer_type Reducer;
+    KernelFunc(NDIt, Reducer);
+
+    typename Reduction::binary_operation BOp;
+    // Compute the partial sum for the work-group.
+    // Then the 0-th work-iteam stores it to the final/global reduction variable.
+    Reducer.MValue = ext::oneapi::reduce(NDIt.get_group(), Reducer.MValue, BOp);
+    if (NDIt.get_local_linear_id() == 0)
+      Reducer.atomic_combine(Reduction::getOutPointer(Out));
+```
+
+The most general case is when the `fast atomics` and `fast reduce` are not available.
+It computes the partial sum using tree-reduction loop and stores the partial sum for the work-group to a global array of partial sums, which later must be also reduced (by using additional(s) kernel(s)). This general case algorithm also requires allocation of a local memory for the tree-reduction loop:
+```c++
+    // Call user's functions. Reducer.MValue gets initialized there.
+    typename Reduction::reducer_type Reducer(ReduIdentity, BOp);
+    KernelFunc(NDIt, Reducer);
+
+    size_t WGSize = NDIt.get_local_range().size();
+    size_t LID = NDIt.get_local_linear_id();
+    // Copy the element to local memory to prepare it for tree-reduction.
+    LocalReds[LID] = Reducer.MValue;
+    if (!IsPow2WG)
+      LocalReds[WGSize] = ReduIdentity;
+    NDIt.barrier();
+
+    // Tree-reduction: reduce the local array LocalReds[:] to LocalReds[0]
+    // LocalReds[WGSize] accumulates last/odd elements when the step
+    // of tree-reduction loop is not even.
+    size_t PrevStep = WGSize;
+    for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+      if (LID < CurStep)
+        LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+      else if (!IsPow2WG && LID == CurStep && (PrevStep & 0x1))
+        LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+      NDIt.barrier();
+      PrevStep = CurStep;
+    }
+
+    // Compute the partial sum/reduction for the work-group and store it to the global array of partial sums.
+    if (LID == 0) {
+      size_t GrID = NDIt.get_group_linear_id();
+      typename Reduction::result_type PSum =
+          IsPow2WG ? LocalReds[0] : BOp(LocalReds[0], LocalReds[WGSize]);
+      if (IsUpdateOfUserVar) // depends on accessor read/write mode or on property::reduction::initialize_to_identity
+        PSum = BOp(*(Reduction::getOutPointer(Out)), PSum);
+      Reduction::getOutPointer(Out)[GrID] = PSum;
+    }
+```
+
+The *additional kernel(s)* that reduce the partial sums look very similar to the *main* kernel, the only difference is that they do not call user's function.
+
+---
+TODO #1 (Performance): After the reduction accepting 'range' was implemented it became possible to simply re-use that implementation for the *additional kernel* runs. Instead of running the *additional kernel* several times (if the number of partial sums is still big), it is better to run the parallel_for accepting 'range' once.
```

Comment:
I think you should qualify here which devices you're talking about, unless these statements about performance apply to all backends and devices.

Reply:
This is the general comment, i.e. for all backends/devices. If there is an efficient implementation of parallel_for(range, reduction), then why not use that to reduce let's say 1M partial sums, instead of inventing/adding N calls of additional kernels inside parallel_for(nd_range, reduction) implementation.